### PR TITLE
Sync remote resources in groups edit

### PR DESCRIPTION
**Pull request by davidcaron**
## Description
- The remote resources are stored in a new table, with a similar structure as 'regular' resources. Another new table stores one record per service with the synchronization information (last sync, service type, service name, service id, etc.)
- The synchronization is done every hour by a cron service, or directly in the UI if requested by the user. The time of the last synchronization is displayed in the UI.
- The resources displayed in the 'group edit' view are either remote or local, the user doesn't need to know. Once a permission is associated with a resource in the UI, the remote resources is written to the regular Resource table.
- The cron service will also remove from the local Resource table every resource that doesn't have any child resources and has no group or user permission associated.
- If a resource is absent from the remote server, the UI displays a flag next to the resource, and a `Clean` button to remove the resource from the database. There is also a `Clean all` button available. 
## Supported services
- Thredds
- Geoserver
- Project-api
## Other notes
There is a direct database call from the UI in this PR. A note has been added in the code.

### CODE REVIEW

**Comment by davidcaron** [comment]
Related to #42 .

**Comment by davidcaron** [comment]
One more thing I noticed on Colibri:
- When the sync is not implemented or it never ran, the resources are flagged as not on the remote server and the 'Clean' button is available.

File: magpie/helpers/sync_resources.py [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+
+    # loop the resource tree by reversed ordering (starting from the leaves)
+    # if the resource doesn't have any children or permissions, delete it
+    for resource in session.query(models.Resource).order_by(models.Resource.ordering.desc()):
+        if resource.resource_type_name == 'service':
+            continue
+        n_children = resource_tree_service.count_children(resource.resource_id, session)
+        if n_children == 0:
+            if not resource.group_permissions and not resource.user_permissions:
+                session.delete(resource)
```
[the remark this answers is not in the document]
Comment (davidcaron) [reply magpie/helpers/sync_resources.py]: @dbyrns I have no problem with keeping or deleting the resource. It shouldn't make any difference to the UI either. We just need to reach a consensus.

**Comment by dbyrns** [comment]
Maybe this could be avoided by checking the last run datetime property. We could replace the clean button by an out-of-sync warning if the last run never occurs or occurs too far in the past.

**Comment by fmigneault** [comment]
@dbyrns which delta would be good to consider 'too far in the past' ?

**Comment by dbyrns** [comment]
I guess something based on the cron job interval... After twice its interval, it means that something is broken and we cannot trust anymore the remote table.

File: magpie/helpers/sync_resources.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dbyrns) [reply magpie/helpers/sync_resources.py]: I got the @fmigneault point. If a remote service create a resource it should safely assumes that it exists in magpie even if no permissions are set.

**Comment by davidcaron** [comment]
@fmigneault @dbyrns 
colibri is updated with this branch. It covers everything we mentioned.

**Comment by dbyrns** [comment]
Some observations:
* It seems like the cron job has not run or is having issues with all the services. There was a banner telling me that.
* I got an error once while trying to synch with thredds (error about reaching the private url I think). On following tests everything was fine.
* Finally the resources for thredds are under Birdhouse with a capital B. But the url will match with birdhouse (all in lowercase). I can confirm looking at thredds that birdhouse has a capital B but we should look for a way to extract the url equivalent of the resource else we will got problems matching the permissions.

**Comment by davidcaron** [comment]
It's up to date on colibri now. 
- For the cron service, I had to get settings from `/opt/local/src/magpie/env/magpie.env` but also from `/opt/local/src/magpie/magpie/env/magpie.env` because docker-compose puts magpie settings there. Docker-compose doesn't actually overwrite `/opt/local/src/magpie/env/magpie.env`. I think these should be merged, I would find it simpler...
- About capital letters in the thredds catalog: the name of the resource can be capitalized
and not its id. But in other cases, both the name and its id are
capitalized. I made sure to get the actual id of the resource which matches the case of
the url.

**Comment by fmigneault** [comment]
@davidcaron 
I would make sure to map both env files in `volumes` of magpie in the docker-compose so they are overridden.
I think that if they are overridden, the cron would retrieve the good values using the `get_constant` method and such.
I believe it is better to override than merge because it is (1) easier to apply modifications, (2) easier to resolve conflicting values (which one has priority?) and (3) easier to debug by knowing where values come from.

File: magpie/ui/management/templates/tree_scripts.mako [hunk 2]
```diff
@@ -42,16 +42,15 @@ li.Collapsed {
     <ul>
      %for key in tree:
         <div class="clear"/>
-        <form id="resource_${tree[key]['id']}" action="${request.path}" method="post">
+        <form id="resource_${tree[key]['id']}_${tree[key].get('remote_id', '')}" action="${request.path}" method="post">
```
Comment (fmigneault) [inline magpie/ui/management/templates/tree_scripts.mako]: Has this been validated to work with other kind of services that do not employ the cron?
Does `render_tree` still work with other pages that might not have `_remoteID` in their id?

**Comment by fmigneault** [comment]
Also depends on https://github.com/Ouranosinc/PAVICS/pull/99/commits/5c33418f5bf6208feddcb951a6cb3df7debfb45e
to obtain magpie env overrides in docker-compose

**Comment by fmigneault** [comment]
@dbyrns  @davidcaron 
Wondering about your opinions on following feature... which impacts this branch.

For OGC, to allow read/write control of any `api` routes combination, I'm trying to modify the `api` service type to make it generic so that it is not implementation specific (ie:  `geoserver-api` and `project-api` would both be of type `api`, not individual types).

This causes the sync service to fail because types `geoserver-api` / `project-api` cannot be found anymore (see image). 
Furthermore, any resource permission POST fails because corresponding resource `type` are not found.

image

**_Therefore, my question is:_**

Should we (1) make service sync dependent of the service `name` rather than its `type`, considering that similar service types do not necessarily behave the same way under the hood (ex: API routes are not the sames) or (2) always add a new derived type from `ServiceAPI` for every single new case we want to protect in the future. 

For (2), I find it makes it a bit pointless to have a `type` since they will always match a unique service of same `name`.

**Comment by dbyrns** [comment]
If I got it, we have multiple services sharing a common url/permission scheme but not sharing their resources retrieval method. Could we had a second interface for the resources retrieval method and upon service creation we choose 1) the service type and 2) the resources type?
This way : 
- multiple service can share the same service / resource type (e.g. multiple thredds instance)
- two services can share the same service type with different resource types (e.g. geoserver / project api)
- two services can share the same resource types with different service type (e.g. [geoserver-api, geoserver wfs, geoserver wms] or [thredds, ncwms])

**Comment by fmigneault** [comment]
Yes that's pretty much the idea of it, and 2 types would be a great solution.
I think this could be done quite easily as `_SyncServiceInterface` derived classes are already applied according to a dict here :
https://github.com/Ouranosinc/Magpie/blob/344ace6c3d584eb7cdff191143588cccf689f131/magpie/helpers/sync_resources.py#L25-L31

We could have a `sync_type = { 'none', 'project-api', 'geoserver-api', ... }` which defines how service resources are updated, although they got the same `service_type`  and underlying `resource_type`.
We would only need to add a column `sync_type` to the service's db table, and query it as needed instead of static dict lookup.

**Comment by fmigneault** [comment]
@davidcaron 
Just noticed this in the code, replace `iteritems()` by `items()` calls to be Py2/3 compatible.

File: magpie/register.py [hunk 3]
```diff
@@ -359,7 +359,8 @@ def magpie_register_services_with_db_session(services_dict, db_session, push_to_
     for svc_name, svc_values in services_dict.items():
         svc_new_url = os.path.expandvars(svc_values['url'])
         svc_type = svc_values['type']
-        svc_sync_type = svc_values['sync_type']
+
+        svc_sync_type = svc_values.get('sync_type', svc_values['title'])
```
Comment (fmigneault) [inline magpie/register.py]: Maybe instead of 'title', let's default to None or other value reflecting not-implemented?

**Comment by fmigneault** [comment]
Routes like `/users/francis/resources` return `"service_sync_type": "None"`
Value should be either `null` (converted by json from `None`) or any corresponding 'default' value.

**Pull request merged**